packages/resource/src/index.ts

```diff
@@ -0,0 +1,237 @@
+ * - `timeout`: time in Milliseconds after which the fetcher aborts automatically
+ * - `noAutoAbort`: can be set to true to make a new source not automatically abort a previous request
+ */
+export function makeAbortable(
+  options?: AbortableOptions,
+): [signal: () => AbortSignal, abort: () => void] {
+  let controller: AbortController | undefined;
+  let timeout: NodeJS.Timeout | number | undefined;
+  const abort = () => {
+    timeout && clearTimeout(timeout);
+    controller?.abort();
+  };
+  const signal = () => {
+    if (!options?.noAutoAbort && controller?.signal.aborted === false) {
+      abort();
+    }
+    controller = new AbortController();
+    if (options?.timeout) {
+      timeout = setTimeout(abort, options.timeout);
+    }
+    return controller.signal;
+  };
+  return [signal, abort];
+}
+
+const mapEntries = (entries: [key: string, value: any][]) =>
+  entries.map(([key, value]) => `${serializer(key)}: ${serializer(value)}`);
+
+/** **Serializes a request source** */
+export const serializer = (req: any): string => {
+  const rtype = typeof req;
+  if (
+    !req ||
+    rtype === "boolean" ||
+    rtype === "number" ||
+    rtype === "bigint" ||
+    rtype === "function"
+  ) {
+    return req + "";
+  } else if (rtype === "string") {
+    return `"${req}"`;
+  } else if (Array.isArray(req)) {
+    return "[" + req.map(serializer).join(", ") + "]";
+  } else if ("entries" in (req || {}) && typeof req.entries === "function") {
+    return "{" + mapEntries([...req.entries()]) + "}";
+  } else if (rtype === "object") {
+    return "{" + mapEntries(Object.entries(req)) + "}";
+  } else {
+    return "";
+  }
+};
+
+export type CacheEntry<T, S> = {
+  source: S;
+  data: T;
+  ts: number;
+};
+export const cache: Record<string, CacheEntry<any, any>> = {};
+
+export type CacheOptions<T, S> = {
+  cache?: Record<string, CacheEntry<T, S>>;
+  expires?: number | ((entry: Omit<CacheEntry<T, S>, "ts">) => number);
+  storage?: Storage;
+  storageKey?: string;
+};
+
+/**
+ * **Creates a caching fetcher**
+ * ```ts
+ * const [fetcher, invalidate, expired] =
+ *   makeCache(
+ *     (source) => fetch(source).then(r => r.json()),
+ *     { storage: localStorage, storageKey: 'foo' }
+ *   );
+ * ```
+ * Wraps the fetcher to use a cache. Returns the wrapped fetcher, an invalidate callback that requires the source to invalidate the request and a signal accessor with the last automatically invalidated request.
+ *
+ * Can be customized with the following optional options:
+ * - `cache` - allows to use a local cache instead of the global one
+ * - `expires` - allows to define a custom timeout; either accepts a number or a function that receives an object with source and data of the request and returns a number in Milliseconds
+ * - `storage` - a storage like localStorage to persist the cache over reloads
+ * - `storageKey` - the key which is used to store the cache in the storage
+ */
+export function makeCache<T, S, R>(
```

Comment:
I don't like that the cache makes so many assumptions.
1. What if the params / result are not, and shouldn't be serializable?
2. What If I want to clear all the cache? not one by one by a single key
3. What if I don't care about expiry data, shouldn't this be just another primitive that hooks into `invalidate` or `invalidateAll`?

Probably doesn't matter now as it's focusing on createResource and fetching, but in the future it could be generalized and extracted to a separate package.

Reply:
1. Then you should replace them with a serializable key, as stated in the documentation. That being said, I'm able to serialize more than JSON.stringify can handle.
2. That's actually a use case I should cover with invalidate.all(), let me add that later.
3. I originally had that separated in the fetch primitive, but the logic to allow for this separation was larger than the signal.

Reply:
That being said: how would you persist the cache at all then?